wit/tls.wit

```diff
@@ -0,0 +1,309 @@
+    /// data from the socket into the `public-output` stream and
+    /// data from the `public-input` into the socket.
+    ///
+    /// # Caution
+    /// Because the guest acts as both the producer and the consumer for these
+    /// streams, do not use the `blocking_*` methods as that will deadlock yourself.
+    record io-streams {
+        public-input: input-stream,
+        public-output: output-stream,
+        private-output: output-stream,
+        private-input: input-stream,
+    }
+
+    flags client-suspension-points {
+        /// When the client received the server's certificate.
+        verify-server-identity,
+
+        /// When the client received a certificate request from the server.
+        select-client-identity,
+
+        /// When the initial handshake was successful.
+        connected,
+    }
+
+    resource client-suspension {
+        /// At which point the TLS stream is suspended.
+        /// Exactly one flag is set in the return value.
+        at: func() -> client-suspension-points;
+
+        /// Only for select-client-identity:
+        // TODO: acceptable-authorities: func() -> result<list<string>>;
+        
+        /// Only for verify-server-identity:
+        // TODO: unverified-identity: func() -> result<public-identity>;
+    }
+
+    enum suspend-error {
+        not-ready,
+        already-suspended,
+        closed,
+    }
+
+
+
+
+
+
+
+
+
+    // TODO
+    resource server {
+        constructor(suspend-at: server-suspension-points);
+        streams: func() -> result<io-streams>;
+
+        configure-alpn-ids: func(value: list<alpn-id>) -> result;
+        configure-identities: func(value: list<borrow<private-identity>>) -> result;
+
+        server-name: func() -> option<string>;
+        alpn-id: func() -> option<alpn-id>;
+        protocol-version: func() -> option<protocol-version>;
+        client-identity: func() -> option<public-identity>;
+        server-identity: func() -> option<private-identity>;
+
+        suspend: func() -> result<server-suspension, suspend-error>;
+        resume: func() -> result;
+        subscribe: func() -> pollable;
+    }
+
+    flags server-suspension-points {
+        /// When the server received the initial message from the client.
+        client-hello,
+
+        /// When the server received the client's certificate.
+        verify-client-identity,
+
+        /// When the initial handshake was successful.
+        accepted,
+    }
+
+    resource server-suspension {
+        at: func() -> server-suspension-points;
+
+        /// Only for client-hello:
+        // TODO: requested-protocol-versions: func() -> result<list<protocol-version>>;
+        // TODO: requested-server-name: func() -> result<option<string>>;
+        // TODO: requested-alpn-ids: func() -> result<list<alpn-id>>;
+
+        /// Only for verify-client-identity:
+        // TODO: unverified-identity: func() -> result<public-identity>;
+    }
+}
```

Comment:
Why don't we flip the abstraction here and make each suspension point a resource in itself? (see suggestion in review comment)